src/client.rs

```diff
@@ -1140,6 +1198,61 @@ impl Client {
     }
 }
 
+/// A client for pushing OCI artifacts to a registry from a stream.
+pub struct AsyncBlobPusher {
+    client: Client,
+    offset: usize,
+    location: String,
+    image: Reference,
+    digest: String,
+    pending: Option<Pin<Box<dyn Future<Output = Result<(String, usize)>> + Send>>>,
+}
+
+impl AsyncWrite for AsyncBlobPusher {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::result::Result<usize, std::io::Error>> {
+        let this = self.get_mut();
+        this.pending = match this.pending.take() {
+            Some(mut pending) => match pending.poll_unpin(cx) {
+                Poll::Ready(_) => return Poll::Ready(Ok(buf.len())),
+                Poll::Pending => Some(pending),
+            },
+            None => Some(
+                this.client
+                    .clone()
+                    .push_chunk_request(
+                        this.location.clone(),
+                        this.image.clone(),
+                        buf.into(),
+                        this.offset,
+                    )
+                    .boxed(),
+            ),
+        };
+        Poll::Pending
+    }
+
+    fn poll_flush(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<std::result::Result<(), std::io::Error>> {
+        match self.pending {
+            Some(_) => Poll::Pending,
```

Comment:
Should this poll_unpin too while pending?

Reply:
Since we're "flushing" inside the `pending` future, as long as there's something there, it's not flushed – that's my thinking